sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.splittabledofn;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.sdk.io.range.ByteKeyRange;
+import org.apache.beam.sdk.transforms.DoFn;
+
+/**
+ * A {@link RestrictionTracker} for claiming {@link ByteKey}s in a {@link ByteKeyRange} in a
+ * monotonically increasing fashion.
+ */
+public class ByteKeyRangeTracker extends RestrictionTracker<ByteKeyRange, ByteKey> {
+  private ByteKeyRange range;
+  @Nullable private ByteKey lastClaimedKey = null;
+  @Nullable private ByteKey lastAttemptedKey = null;
+
+  private ByteKeyRangeTracker(ByteKeyRange range) {
+    this.range = checkNotNull(range);
+  }
+
+  /**
+   * Instantiates a new {@link ByteKeyRangeTracker} with the specified range. The keys in the range
+   * are left padded to be the same length in bytes.
+   */
+  public static ByteKeyRangeTracker of(ByteKeyRange range) {
+    return new ByteKeyRangeTracker(ByteKeyRange.of(range.getStartKey(), range.getEndKey()));
+  }
+
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey, range);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
+    // must always be less than *end*.
+    ByteKey rangeEndKey =
+        (nextKey.equals(range.getEndKey())) ? next(range.getEndKey(), range) : range.getEndKey();
+    ByteKeyRange res = ByteKeyRange.of(nextKey, rangeEndKey);
+    this.range = ByteKeyRange.of(range.getStartKey(), nextKey);
```

Comment:
Good remark, but I am not sure if the extra check is needed, I think it is already covered by the tryClaim assertions. If you arrive here the lastClaimedKey must be consistent or it will fail before. Probably that is more clear in the tests.